flex_model/core/wrapper.py

```diff
@@ -59,13 +61,170 @@ class _SharedState:
     offload_mode: str
 
 
+class _HookFunctionGroupManager:
+    def __init__(self):
+        self.hook_fn_to_groups_map = defaultdict(set)
+        self.groups = set()
+
+    def get_hook_fn_groups(self, hook_fn: HookFunction) -> Set[str]:
+        return self.hook_fn_to_groups_map[hook_fn]
+
+    def get_group_hook_fns(self, group_name: str) -> List[HookFunction]:
+        hook_fns = []
+        for hook_fn, groups in self.hook_fn_to_groups_map.items():
+            if group_name in groups:
+                hook_fns.append(hook_fn)
+        return hook_fns
+
+    def create(
+        self,
+        group_name: str,
+        group_constructor: str,
+        expected_shape: Tuple[Optional[int], ...],
+        all_names: List[str],
+        editing_function: Optional[Callable] = None,
+        unpack_idx: int = 0,
+    ) -> List[HookFunction]:
+        """Instantiate and group hook functions based on module name pattern matching."""
+        matching_modules = []
+        for module_name in all_names:
+            if group_constructor in module_name:
+                matching_modules.append(module_name)
+
+        hook_fns = []
+        for mod_name in matching_modules:
+            hf = HookFunction(
+                module_name=mod_name,
+                expected_shape=expected_shape,
+                editing_function=editing_function,
+                unpack_idx=unpack_idx,
+            )
+            self.hook_fn_to_groups_map[hf].add(group_name)
+            hook_fns.append(hf)
+
+        if group_name not in self.groups:
```

Comment:
What is the functionality if this group already exists?

Reply:
For example, if I wanted to create a group of hook functions whose submodule runs a matrix multiplication, I could call `create` twice, once for modules matching the `self_attn` string and other matching the `fc1` or `fc2` string. These would have the same `matmul` group tag.